executor/common_linux.h

```diff
@@ -4535,3 +4613,280 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+	if (props->mac)
+		netlink_attr(nlmsg, NL80211_ATTR_MAC, props->mac, ETH_ALEN);
+	if (props->wiphy_freq_fixed)
+		netlink_attr(nlmsg, NL80211_ATTR_FREQ_FIXED, NULL, 0);
+	int err = netlink_send(nlmsg, sock);
+	if (err < 0) {
+		debug("join_ibss failed: %s\n", strerror(err));
+		return -1;
+	}
+	return 0;
+}
+
+static int get_ifla_operstate(struct nlmsg* nlmsg, const char* interface_name)
+{
+	struct ifinfomsg info;
+	memset(&info, 0, sizeof(info));
+	info.ifi_family = AF_UNSPEC;
+	info.ifi_index = if_nametoindex(interface_name);
+
+	int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (sock == -1) {
+		debug("get_ifla_operstate: socket failed: %d\n", errno);
+		return -1;
+	}
+
+	netlink_init(nlmsg, RTM_GETLINK, 0, &info, sizeof(info));
+	int n;
+	int err = netlink_send_ext(nlmsg, sock, RTM_NEWLINK, &n);
+	close(sock);
+
+	if (err) {
+		debug("get_ifla_operstate: failed to query: %s\n", strerror(err));
+		return -1;
+	}
+
+	struct rtattr* attr = IFLA_RTA(NLMSG_DATA(nlmsg->buf));
+	for (; RTA_OK(attr, n); attr = RTA_NEXT(attr, n)) {
+		if (attr->rta_type == IFLA_OPERSTATE)
+			return *((int32_t*)RTA_DATA(attr));
+	}
+
+	return -1;
+}
+
+static long syz_80211_join_ibss(volatile long a0, volatile long a1, volatile long a2, volatile long a3)
+{
+	uint8* ssid = (uint8*)a0;
+	int ssid_len = (int)a1;
+	int fix_frequency = (int)a2; // This parameter essentially determines whether it will perform a scan
+	int await_up = (int)a3;
+	struct nlmsg tmp_msg;
+	uint8 bssid[ETH_ALEN] = WIFI_IBSS_BSSID;
+
+	if (ssid == NULL || ssid_len < 0 || ssid_len > WIFI_MAX_SSID_LEN)
+		return -1; // restricting input in order to avoid unnecessary segfaults on memcpy
+
+	int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC);
+	if (sock < 0) {
+		debug("syz_80211_join_ibss: socket creation failed, errno %d\n", errno);
+		return -1;
+	}
+
+	int nl80211_family_id = netlink_query_family_id(&tmp_msg, sock, "nl80211");
+	struct join_ibss_props ibss_props = {
+	    .wiphy_freq = WIFI_DEFAULT_FREQUENCY, .wiphy_freq_fixed = (fix_frequency > 0), .mac = bssid, .ssid = ssid, .ssid_len = ssid_len};
+
+	for (int device_id = 0; device_id < WIFI_INITIAL_DEVICE_COUNT; device_id++) {
+		char interface[6] = {0};
+		memcpy(interface, "wlan0", 5);
+		interface[4] += device_id;
+
+		int ret = nl80211_set_interface(&tmp_msg, sock, nl80211_family_id, interface, NL80211_IFTYPE_ADHOC);
+		if (ret < 0) {
+			debug("syz_80211_join_ibss: nl80211_set_interface failed for #%d, ret %d\n", device_id, ret);
+			goto error;
+		}
+
+		ret = set_interface_state(interface, 1);
+		if (ret < 0) {
+			debug("set_interface_state: set_interface_state failed for #%d, ret %d\n", device_id, ret);
+			goto error;
+		}
+
+		ret = nl80211_join_ibss(&tmp_msg, sock, nl80211_family_id, interface, &ibss_props);
+		if (ret < 0) {
+			debug("syz_80211_join_ibss: nl80211_join_ibss failed for #%d, ret %d\n", device_id, ret);
+			goto error;
+		}
+
+		if (await_up) {
```

Comment:
Hm, can we ask all devices to join the network first, and then wait until they do? Instead of doing it one by one.

Reply:
I'm worried that this additional flexibility can complicate generation of good progs. I've implemented it in a separate branch  and starter a separate syzkaller instance. If it works fine, I'll push that commit into this pull request.

Reply:
Syzkaller from this pull request detected 12 types of crashes during 3 days, while syzkaller with `syz_80211_join_ibss(wlan_interface, ssid, ssid_len, fix_requency, await_up)`  detected only 7 types of crashes during the same time period. Of course, this can also be attributed to (un)luck as the process is somewhat random, but still the difference is sufficient.

Reply:
Oh, I posted my replies into the wrong comment. They were in repsonse to `This pseudo-syscall still seems suspicious to me. Would it make sense to change it to only make one device join the network instead of all of them?`